<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: ATAC-seq Quality Control</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> ATAC-seq Quality Control
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;ATACseqQC&rsquo; version 1.16.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="ATACseqQC-package.html">ATACseqQC-package</a></td>
<td>ATAC-seq Quality Control</td></tr>
<tr><td style="width: 25%;"><a href="ATACseqQC-package.html">ATACseqQC</a></td>
<td>ATAC-seq Quality Control</td></tr>
<tr><td style="width: 25%;"><a href="bamQC.html">bamQC</a></td>
<td>Mapping quality control</td></tr>
<tr><td style="width: 25%;"><a href="DB.html">DB</a></td>
<td>helper function for differential binding</td></tr>
<tr><td style="width: 25%;"><a href="distanceDyad.html">distanceDyad</a></td>
<td>Distance of potential nucleosome dyad</td></tr>
<tr><td style="width: 25%;"><a href="enrichedFragments.html">enrichedFragments</a></td>
<td>enrichment for nucleosome-free fragments and nucleosome signals</td></tr>
<tr><td style="width: 25%;"><a href="estimateLibComplexity.html">estimateLibComplexity</a></td>
<td>Library complexity estimation</td></tr>
<tr><td style="width: 25%;"><a href="factorFootprints.html">factorFootprints</a></td>
<td>plot ATAC-seq footprints infer factor occupancy genome wide</td></tr>
<tr><td style="width: 25%;"><a href="footprintsScanner.html">footprintsScanner</a></td>
<td>scan ATAC-seq footprints infer factor occupancy genome wide</td></tr>
<tr><td style="width: 25%;"><a href="fragSizeDist.html">fragSizeDist</a></td>
<td>fragment size distribution</td></tr>
<tr><td style="width: 25%;"><a href="NFRscore.html">NFRscore</a></td>
<td>Nucleosome Free Regions (NFR) score</td></tr>
<tr><td style="width: 25%;"><a href="peakdet.html">peakdet</a></td>
<td>Detect peak positions</td></tr>
<tr><td style="width: 25%;"><a href="plotCorrelation.html">plotCorrelation</a></td>
<td>plot Correlations of multiple samples</td></tr>
<tr><td style="width: 25%;"><a href="plotFootprints.html">plotFootprints</a></td>
<td>Plots a footprint estimated by Centipede</td></tr>
<tr><td style="width: 25%;"><a href="footprintsScanner.html">prepareBindingSitesList</a></td>
<td>scan ATAC-seq footprints infer factor occupancy genome wide</td></tr>
<tr><td style="width: 25%;"><a href="PTscore.html">PTscore</a></td>
<td>Promoter/Transcript body (PT) score</td></tr>
<tr><td style="width: 25%;"><a href="pwmscores.html">pwmscores</a></td>
<td>max PWM scores for sequences</td></tr>
<tr><td style="width: 25%;"><a href="readBamFile.html">readBamFile</a></td>
<td>read in bam files</td></tr>
<tr><td style="width: 25%;"><a href="readsDupFreq.html">readsDupFreq</a></td>
<td>Calculating duplication frequency</td></tr>
<tr><td style="width: 25%;"><a href="saturationPlot.html">saturationPlot</a></td>
<td>Plotting Saturation curves</td></tr>
<tr><td style="width: 25%;"><a href="shiftGAlignments.html">shiftGAlignments</a></td>
<td>shift 5' ends for single end reads</td></tr>
<tr><td style="width: 25%;"><a href="shiftGAlignmentsList.html">shiftGAlignmentsList</a></td>
<td>shift 5' ends</td></tr>
<tr><td style="width: 25%;"><a href="shiftReads.html">shiftReads</a></td>
<td>shift read for 5'end</td></tr>
<tr><td style="width: 25%;"><a href="splitBam.html">splitBam</a></td>
<td>prepare bam files for downstream analysis</td></tr>
<tr><td style="width: 25%;"><a href="splitGAlignmentsByCut.html">splitGAlignmentsByCut</a></td>
<td>split bams into nucleosome free, mononucleosome, dinucleosome and trinucleosome</td></tr>
<tr><td style="width: 25%;"><a href="TSSEscore.html">TSSEscore</a></td>
<td>Transcription Start Site (TSS) Enrichment Score</td></tr>
<tr><td style="width: 25%;"><a href="vPlot.html">vPlot</a></td>
<td>V-plot</td></tr>
<tr><td style="width: 25%;"><a href="writeListOfGAlignments.html">writeListOfGAlignments</a></td>
<td>export list of GAlignments into bam files</td></tr>
</table>
</body></html>
