<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: enrichment for nucleosome-free fragments and nucleosome...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for enrichedFragments {ATACseqQC}"><tr><td>enrichedFragments {ATACseqQC}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>enrichment for nucleosome-free fragments and nucleosome signals</h2>

<h3>Description</h3>

<p>Get the enrichment signals for nucleosome-free fagments and
nucleosomes.
</p>


<h3>Usage</h3>

<pre>
enrichedFragments(
  bamfiles,
  index = bamfiles,
  gal,
  TSS,
  librarySize,
  upstream = 1010L,
  downstream = 1010L,
  n.tile = 101L,
  normal.method = "quantile",
  adjustFragmentLength = 80L,
  TSS.filter = 0.5,
  seqlev = paste0("chr", c(1:22, "X", "Y"))
)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>bamfiles</code></td>
<td>
<p>A vector of characters indicates the file names of bams.</p>
</td></tr>
<tr valign="top"><td><code>index</code></td>
<td>
<p>The names of the index file of the 'BAM' file being processed;
This is given without the '.bai' extension.</p>
</td></tr>
<tr valign="top"><td><code>gal</code></td>
<td>
<p>A GAlignmentsList object or a list of GAlignmentPairs.
If bamfiles is missing, gal is required.</p>
</td></tr>
<tr valign="top"><td><code>TSS</code></td>
<td>
<p>an object of <a href="../../GenomicRanges/help/GRanges-class.html">GRanges</a> indicates
the transcript start sites. All the width of TSS should equal to 1.
Otherwise, TSS will be reset to the center of input TSS.</p>
</td></tr>
<tr valign="top"><td><code>librarySize</code></td>
<td>
<p>A vector of numeric indicates the library size. Output of
<a href="../../ChIPpeakAnno/help/estLibSize.html">estLibSize</a></p>
</td></tr>
<tr valign="top"><td><code>upstream, downstream</code></td>
<td>
<p>numeric(1) or integer(1).
Upstream and downstream size from each TSS.</p>
</td></tr>
<tr valign="top"><td><code>n.tile</code></td>
<td>
<p>numeric(1) or integer(1). The number of tiles to generate
for each element of TSS.</p>
</td></tr>
<tr valign="top"><td><code>normal.method</code></td>
<td>
<p>character(1). Normalization methods,
could be &quot;none&quot; or &quot;quantile&quot;.
See <a href="../../limma/help/normalizeBetweenArrays.html">normalizeBetweenArrays</a>.</p>
</td></tr>
<tr valign="top"><td><code>adjustFragmentLength</code></td>
<td>
<p>numeric(1) or integer(1).
The size of fragment to be adjusted to.
Default is set to half of the nucleosome size (80)</p>
</td></tr>
<tr valign="top"><td><code>TSS.filter</code></td>
<td>
<p>numeric(1). The filter for signal strength of each TSS.
Default 0.5 indicates the average signal strength for the TSS
from upstream to downstream bins should be greater than 0.5.</p>
</td></tr>
<tr valign="top"><td><code>seqlev</code></td>
<td>
<p>A vector of character indicates the sequence names to be
considered.</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A list of matrixes. In each matrix, each row record the signals for
corresponding feature.
</p>


<h3>Author(s)</h3>

<p>Jianhong Ou
</p>


<h3>Examples</h3>

<pre>

bamfiles &lt;- system.file("extdata", "splited",
                       c("NucleosomeFree.bam",
                         "mononucleosome.bam",
                         "dinucleosome.bam",
                         "trinucleosome.bam"), package="ATACseqQC")
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txs &lt;- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
TSS &lt;- promoters(txs, upstream=0, downstream=1)
library(ChIPpeakAnno)
librarySize &lt;- estLibSize(bamfiles)
sigs &lt;- enrichedFragments(bamfiles, TSS=TSS, librarySize=librarySize,
                          seqlev="chr1", TSS.filter=0)
sigs.log2 &lt;- lapply(sigs, function(.ele) log2(.ele+1))
featureAlignedHeatmap(sigs.log2, reCenterPeaks(TSS, width=2020),
                      zeroAt=.5, n.tile=101, upper.extreme=2)
featureAlignedDistribution(sigs, reCenterPeaks(TSS, width=2020),
                           zeroAt=.5, n.tile=101, type="l")

</pre>

<hr /><div style="text-align: center;">[Package <em>ATACseqQC</em> version 1.16.0 <a href="00Index.html">Index</a>]</div>
</body></html>
