<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: scan ATAC-seq footprints infer factor occupancy genome wide</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for footprintsScanner {ATACseqQC}"><tr><td>footprintsScanner {ATACseqQC}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>scan ATAC-seq footprints infer factor occupancy genome wide</h2>

<h3>Description</h3>

<p>Aggregate ATAC-seq footprint for a bunch of motifs generated
over binding sites within the genome.
</p>


<h3>Usage</h3>

<pre>
footprintsScanner(
  bamExp,
  bamCtl,
  indexExp = bamExp,
  indexCtl = bamCtl,
  bindingSitesList,
  seqlev = paste0("chr", c(1:25, "X", "Y")),
  proximal = 40L,
  distal = proximal,
  gap = 10L,
  maximalBindingWidth = NA,
  cutoffLogFC = log2(1.5),
  cutoffPValue = 0.05,
  correlatedFactorCutoff = 3/4
)

prepareBindingSitesList(
  pfms,
  genome,
  seqlev = paste0("chr", c(1:22, "X", "Y")),
  expSiteNum = 5000
)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>bamExp</code></td>
<td>
<p>A vector of characters indicates the file names of experiment bams. 
The bam file must be the one with shifted reads.</p>
</td></tr>
<tr valign="top"><td><code>bamCtl</code></td>
<td>
<p>A vector of characters indicates the file names of control bams. 
The bam file must be the one with shifted reads.</p>
</td></tr>
<tr valign="top"><td><code>indexExp, indexCtl</code></td>
<td>
<p>The names of the index file of the 'BAM' file being processed;
This is given without the '.bai' extension.</p>
</td></tr>
<tr valign="top"><td><code>bindingSitesList</code></td>
<td>
<p>A object of <a href="../../GenomicRanges/help/GRangesList-class.html">GRangesList</a> indicates
candidate binding sites (eg. the output of fimo).</p>
</td></tr>
<tr valign="top"><td><code>seqlev</code></td>
<td>
<p>A vector of characters indicates the sequence levels.</p>
</td></tr>
<tr valign="top"><td><code>proximal, distal</code></td>
<td>
<p>numeric(1) or integer(1).
basepair for open region from binding sites (proximal) and extented region for background (distal) 
of the binding region for aggregate ATAC-seq footprint.</p>
</td></tr>
<tr valign="top"><td><code>gap</code></td>
<td>
<p>numeric(1) or integer(1). basepair for gaps among binding sites, 
proximal, and distal. default is 5L.</p>
</td></tr>
<tr valign="top"><td><code>maximalBindingWidth</code></td>
<td>
<p>numeric(1) or integer(1). Maximal binding sites width for
all the motifs. If setted, all motif binding sites will be 
re-sized to this value.</p>
</td></tr>
<tr valign="top"><td><code>cutoffLogFC, cutoffPValue</code></td>
<td>
<p>numeric(1). Cutoff value for differential bindings.</p>
</td></tr>
<tr valign="top"><td><code>correlatedFactorCutoff</code></td>
<td>
<p>numeric(1). Cutoff value for correlated factors. 
If the overlapping binding site within 100bp is more than cutoff, the TFs will be treated as correlated factors.</p>
</td></tr>
<tr valign="top"><td><code>pfms</code></td>
<td>
<p>A list of Position frequency Matrix represented as a numeric matrix
with row names A, C, G and T.</p>
</td></tr>
<tr valign="top"><td><code>genome</code></td>
<td>
<p>An object of <a href="../../BSgenome/help/BSgenome-class.html">BSgenome</a>.</p>
</td></tr>
<tr valign="top"><td><code>expSiteNum</code></td>
<td>
<p>numeric(1). Expect number of predicted binding sites.
if predicted binding sites is more than this number, top expSiteNum binding
sites will be used.</p>
</td></tr>
</table>


<h3>Value</h3>

<p>a list.
It includes:
- bindingSites            GRanges of binding site with hits of reads
- data                    a list with test result for each binding site
- results                 a data.frame with open score and enrichment score of motifs
</p>


<h3>Author(s)</h3>

<p>Jianhong Ou
</p>


<h3>Examples</h3>

<pre>

bamfile &lt;- system.file("extdata", "GL1.bam",
                       package="ATACseqQC")
bsl &lt;- system.file("extdata", "jolma2013.motifs.bindingList.95.rds",
                  package="ATACseqQC")
bindingSitesList &lt;- readRDS(bsl)
footprintsScanner(bamfile, seqlev="chr1", bindingSitesList=bindingSitesList)

library(MotifDb)
motifs &lt;- query(MotifDb, c("Hsapiens"))
motifs &lt;- as.list(motifs)
library(BSgenome.Hsapiens.UCSC.hg19)
#bindingSitesList &lt;- prepareBindingSitesList(motifs, genome=Hsapiens)
</pre>

<hr /><div style="text-align: center;">[Package <em>ATACseqQC</em> version 1.16.0 <a href="00Index.html">Index</a>]</div>
</body></html>
